8265123: Add static factory methods to com.sun.net.httpserver.Filter

FrauBoes opened the pull request:
Add two static factory methods to com.sun.net.httpserver.Filter that facilitate the creation of pre- and post-processing Filters: 

`public static Filter beforeResponse(String description, Consumer<HttpExchange> filterImpl) {}`
`public static Filter afterResponse(String description, Consumer<HttpExchange> filterImpl) {}`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265123](https://bugs.openjdk.java.net/browse/JDK-8265123): Add static factory methods to com.sun.net.httpserver.Filter


### Reviewers
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to 8751f6544dd41466138cf650a02726b187649dcc
 * [Michael McMahon](https://openjdk.java.net/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to 0e76acd4fa73a7e59611bd5260bfc0d74f32fb8a
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)


### Contributors
 * Michael McMahon `<michaelm@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3468/head:pull/3468` \
`$ git checkout pull/3468`

Update a local copy of the PR: \
`$ git checkout pull/3468` \
`$ git pull https://git.openjdk.java.net/jdk pull/3468/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3468`

View PR using the GUI difftool: \
`$ git pr show -t 3468`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3468.diff">https://git.openjdk.java.net/jdk/pull/3468.diff</a>

</details>

ChrisHegarty commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -140,4 +142,65 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     * <p>The {@link Consumer filterImpl} is the effective implementation of
+     * the returned Filter and is executed for each {@code HttpExchange} before
+     * the exchange is handled further.
+     *
```
Please expand the description to include similar: "before invoking the next filter in the chain, or the exchange handler (if this is the final filter in the chain)".   Similar comment for afterResponse.

Maybe rename the Consumer do something more general ("consumer") - as the number of references to the term "filter" make it a little hard to follow.

FrauBoes commented:
/csr needed

FrauBoes commented:
/contributor add michaelm

RogerRiggs commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,214 @@
+            logger.setLevel(Level.ALL);
+            ch.setLevel(Level.ALL);
+            logger.addHandler(ch);
+        }
```
Is there a concern that the Logger will be GC'd before it will be used?
Since the reference is not stored anywhere, it can be GC'd immediately.

FrauBoes commented:
Updated the method names and specifications, which now include @apiNotes and examples. I'm not sure if we want to keep both examples for `afterHandler`? 
Added a test for the uncommon case where the filters handle the exchange (not the handler).

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -140,4 +142,117 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     * before {@link Filter.Chain#doFilter(HttpExchange)} is invoked, so before
+     * any subsequent filters in the chain and the exchange handler are
+     * executed. The response is commonly sent by the exchange handler.
+     * The filter is only expected to send the response in the uncommon case
+     * that the exchange will not be handled after the filter is executed.
```
Maybe this could be rephrased as: 

```
 * The filter {@code operation} is typically not expected to handle the request or 
 * {@linkplain HttpExchange#sendResponseHeaders(int, long) send response headers}, since
 * the next filter in the chain (or the exchange handler) will eventually be invoked after the 
 * operation completes.
```

Alternatively we could use a `Predicate<HttpExchange>` instead of a consumer to tell whether the next filter in the chain should be invoked - but that would be another kind of worms. I believe `Consumer<HttpExchange>` is the API we want.

FrauBoes commented:
> This looks a little odd - writing to the output steam before the response headers have been sent. Does it work because the length is actually 0? Please add some response body to the test, this should then show that writing to the output stream before the headers have been sent is an error.

Oops, that's fixed now!

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -140,4 +142,117 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     * rather than modifying it. The filter operation is executed after
+     * {@link Filter.Chain#doFilter(HttpExchange)} is invoked, this means any
+     * subsequent filters in the chain and the exchange handler have been
+     * executed and the response has commonly been sent. The filter is only
+     * expected to send the response in the uncommon case that the exchange
+     * has not been handled before the filter is executed.
```
Maybe we could rephrase it here too:

```
 * executed and the response has commonly been sent. The filter 
 * {@code operation} is typically not expected to handle the exchange or 
 * {@linkplain HttpExchange#sendResponseHeaders(int, long) send the response headers}
 * as doing so is likely to fail, since the exchange is expected to have already been handled 
 * at this point.
```

I would like to make it more obvious that `afterHandler` and `beforeHandler` are not the right API to handle the exchange. I have the feeling that saying that "it is only expected to send ... " does not convey that strongly enough.

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -140,4 +142,117 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     *
+     * <p> Example of adding a sequence of afterHandler filters to a context
+     * (note the ordering):
+     * <pre>{@code
+     *     var filter1 = Filter.afterHandler("Set a1", (var e) -> e.setAttribute("a1", "some value"));
+     *     var filter2 = Filter.afterHandler("Get a1", (var e) -> doSomething(e.getAttribute("a1")));
+     *     var server = HttpServer.create(new InetSocketAddress(0), 10);
+     *     var context = server.createContext("/", new SomeHandler());
+     *     context.getFilters().add(filter2);
+     *     context.getFilters().add(filter1);
+     *     server.start();
+     * }</pre>
```
I don't think "note the ordering" is sufficiently explanatory here: I feel some better description of the sequence of calls that explain that filter2's operation will be executed after filter1's if registered before would be in order. Maybe something like:

```
* <p> Example of adding a sequence of afterHandler filters to a context.<br>
* Because the filter operation is invoked after the next filter in the chain has been invoked,
* the operation of the first after filter registered will be invoked <b>after</b> the operation
* of the second after filter registered (note the ordering):  
* <pre>...</pre>
* <p>The operation of {@code filter2} which has been registered before {@code filter1}
* will be invoked after {@code filter1} operation has been invoked.
```

Maybe renaming `filter1` into `a1Get` and `filter2` into `a1Set` would be less confusing too.

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+        var handler = new EchoHandler();
+        var filter = Filter.beforeHandler("Add x-foo response header",
+                (var e) -> e.getResponseHeaders().set("x-foo", "bar"));
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback please?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+                        throw new UncheckedIOException(ioe);
+                    }
+                });
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+                (var e) -> e.getResponseHeaders().set("x-foo", "bar"));
+        var filter2 = Filter.beforeHandler("Update x-foo response header",
+                (var e) -> e.getResponseHeaders().set("x-foo", "barbar"));
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+        var respCode = new AtomicInteger();
+        var filter = Filter.afterHandler("Log response code",
+                (var e) -> respCode.set(e.getResponseCode()));
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+                (var e) -> e.setAttribute("test-attr", value));
+        var filter2 = Filter.afterHandler("Read attribute",
+                (var e) -> attr.set((String) e.getAttribute("test-attr")));
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+                        throw new UncheckedIOException(ioe);
+                    }
+                });
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+                (var e) -> e.getResponseHeaders().set("x-foo", "bar"));
+        var afterFilter = Filter.afterHandler("Log response code",
+                (var e) -> respCode.set(e.getResponseCode()));
+        var server = HttpServer.create(new InetSocketAddress(0), 10);
```
Can we use the loopback?

dfuch commented on test/jdk/com/sun/net/httpserver/FilterTest.java:
```diff
@@ -0,0 +1,287 @@
+            try (InputStream is = exchange.getRequestBody();
+                 OutputStream os = exchange.getResponseBody()) {
+                var len = is.transferTo(os);
+                exchange.sendResponseHeaders(200, len);
```
Technically this should be: 

```
exchange.sendResponseHeaders(200, len == 0 ? -1 : len);
```

FrauBoes commented:
Thanks for your comments, Daniel. I incorporated them and tightened the specs and examples.

API docs updated in place: http://cr.openjdk.java.net/~jboes/webrevs/8265123/api/jdk.httpserver/com/sun/net/httpserver/Filter.html

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -146,28 +146,26 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     * since this is commonly done by the exchange handler.
      *
-     * <p> Example of adding a response header to all responses:
+     * <p> Example of adding the Foo response header to all responses:
```
Maybe `the {@code "Foo"}` ? Or `the {@code "Foo: Bar"}`?

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -197,41 +195,38 @@ public String description() {
-     * has not been handled before the filter is executed.
+     * executed. The filter {@code operation} is not expected to handle the
+     * exchange or {@linkplain HttpExchange#sendResponseHeaders(int, long) send the response headers}.
+     * Doing so is likely to fail, since this is commonly done by the exchange
+     * handler.
```
maybe `... since the request is expected to have already been  handled before the operation is executed`?

dfuch commented on src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java:
```diff
@@ -197,41 +195,38 @@ public String description() {
-     * <p> Example of adding a sequence of afterHandler filters to a context
-     * (note the ordering):
+     * <p> Example of adding a sequence of afterHandler filters to a context:<br>
+     * The order in which the filters are invoked is reverse to the order in
+     * which they are added to the context's filter-list.
```
May need to be a bit more precise. The filters are invoked in the order they are added, but their operations are invoked in the reverse order. I'd suggest:

```
     * The order in which the filter's operations are invoked is reverse to the order in
     * which the filters are added to the context's filter-list.
```

dfuch reviewed (approved):
I wonder if instead of simply saying:

```
* @return a filter
```

we should say:

```
* @return a filter whose operation is invoked before the exchange is handled
* @return a filter whose operation is invoked after the exchange is handled
```

Otherwise, LGTM!

No need for a new review if you take on these changes.

FrauBoes commented:
/integrate
